File: cli/characterize.py
#!/usr/bin/env python3
"""Behavior-snapshot capture/diff for brownfield work (stdlib-only).

Characterization snapshots pin what code does TODAY (including bugs)
before any edit, so post-change diffs distinguish intended behavior
change from accidental regression. Report-only by design: changed
snapshots warn (review at merge-approval), tool errors fail.

Snapshots live in `./.shiploom/characterization/<name>.json`:
{name, command, exit, outputSha, tail, capturedAt, actor}.
"""

import difflib
import hashlib
import json
import re
from pathlib import Path

from cli import auditlog
from cli import manifest as manifest_mod

SNAPSHOT_DIRNAME = "characterization"
TAIL_LIMIT = 20000
DIFF_CONTEXT = 3
DIFF_MAX_LINES = 60
DEFAULT_TIMEOUT_S = 600

_NAME_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")


def snapshot_dir(project_dir):
    return Path(project_dir) / ".shiploom" / SNAPSHOT_DIRNAME


def _check_name(name):
    if not isinstance(name, str) or not _NAME_RE.match(name):
        raise ValueError("bad snapshot name %r (letters/digits/._-, max 64)" % (name,))


def _resolve_command(project_dir, command, timeout_s):
    """Explicit command wins, else the configured/auto test command."""
    if command:
        return command, timeout_s
    from cli import gates as gates_mod
    configured = gates_mod._configured(Path(project_dir))
    resolved, reason = gates_mod._gate_command("test", configured, project_dir)
    if resolved is None:
        raise ValueError("no command: pass --command or configure gates.test (%s)" % reason)
    entry = configured.get("test") or {}
    timeout = entry.get("timeoutS", DEFAULT_TIMEOUT_S) if isinstance(entry, dict) else DEFAULT_TIMEOUT_S
    return resolved, timeout_s if timeout_s else timeout


def _run(project_dir, command, timeout_s):
    from cli import gates as gates_mod
    return gates_mod._run_command(command, project_dir, timeout_s)


def _sha(text):
    return "sha256:" + hashlib.sha256(text.encode("utf-8")).hexdigest()


def capture(project_dir, name, command=None, timeout_s=0, actor="human"):
    """Run command and store the snapshot. Returns the entry. Raises ValueError."""
    _check_name(name)
    root = Path(project_dir)
    command, timeout = _resolve_command(root, command, timeout_s or 0)
    timeout = timeout or DEFAULT_TIMEOUT_S
    result = _run(root, command, timeout)
    out = result.get("tail", "")
    entry = {"name": name, "command": command, "exit": result.get("exit"),
             "outputSha": _sha(out), "tail": out[-TAIL_LIMIT:],
             "timeoutS": timeout,
             "capturedAt": manifest_mod.utcnow(), "actor": actor}
    dest = snapshot_dir(root)
    dest.mkdir(parents=True, exist_ok=True)
    (dest / ("%s.json" % name)).write_text(
        json.dumps(entry, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    auditlog.append(root, actor=actor, action="characterize.capture", target=name)
    return entry


def _load(project_dir, name):
    _check_name(name)
    path = snapshot_dir(project_dir) / ("%s.json" % name)
    try:
        doc = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError) as exc:
        raise ValueError("no snapshot %r (capture first): %s" % (name, exc))
    if not isinstance(doc, dict) or "outputSha" not in doc:
        raise ValueError("corrupt snapshot %r (recapture)" % name)
    return doc


def list_snapshots(project_dir):
    """Sorted snapshot names present in the project."""
    dest = snapshot_dir(project_dir)
    if not dest.is_dir():
        return []
    return sorted(p.stem for p in dest.glob("*.json") if p.is_file())


def diff(project_dir, name, timeout_s=0):
    """Re-run a snapshot's command and compare. Never raises for run outcomes.

    Returns {"name", "changed", "exitChanged", "outputChanged",
             "unifiedDiff": [...], "current": {exit, outputSha}} or
    {"name", "error": ...} when the comparison itself cannot run.
    Changed behavior is data (warn), not failure: fixes legitimately
    change behavior; humans review diffs at merge-approval.
    """
    try:
        entry = _load(project_dir, name)
    except ValueError as exc:
        return {"name": name, "error": str(exc)}
    root = Path(project_dir)
    try:
        result = _run(root, entry["command"],
                      timeout_s or entry.get("timeoutS") or DEFAULT_TIMEOUT_S)
    except Exception as exc:  # never crash a report on runner faults
        return {"name": name, "error": "runner fault: %s" % exc}
    out = result.get("tail", "")
    exit_changed = result.get("exit") != entry.get("exit")
    output_changed = _sha(out) != entry.get("outputSha")
    unified = []
    if output_changed:
        old = (entry.get("tail") or "").splitlines()
        new = out.splitlines()
        unified = list(difflib.unified_diff(old, new, "before", "after",
                                            n=DIFF_CONTEXT))[:DIFF_MAX_LINES]
    try:
        auditlog.append(root, actor="system", action="characterize.diff", target=name)
    except (OSError, ValueError):
        pass  # reporting must not fail for audit faults; run logs its own trail
    return {"name": name, "changed": bool(exit_changed or output_changed),
            "exitChanged": bool(exit_changed), "outputChanged": bool(output_changed),
            "unifiedDiff": unified,
            "current": {"exit": result.get("exit"), "outputSha": _sha(out)}}

File: internal/characterize/characterize.go
// Package characterize ports cli/characterize.py (stdlib only):
// behavior-snapshot capture/diff for brownfield work. Report-only by
// design: changed snapshots warn, tool errors fail.
package characterize

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/shiploom/ai-builder/internal/auditlog"
	"github.com/shiploom/ai-builder/internal/difflib"
	"github.com/shiploom/ai-builder/internal/execrun"
	"github.com/shiploom/ai-builder/internal/jsoncanon"
	"github.com/shiploom/ai-builder/internal/manifest"
)

// SnapshotDirname is the snapshot directory under .shiploom/.
const SnapshotDirname = "characterization"

// TailLimit mirrors characterize.TAIL_LIMIT (runes).
const TailLimit = 20000

// DiffContext mirrors DIFF_CONTEXT.
const DiffContext = 3

// DiffMaxLines mirrors DIFF_MAX_LINES.
const DiffMaxLines = 60

// DefaultTimeoutS mirrors characterize.DEFAULT_TIMEOUT_S.
const DefaultTimeoutS = 600

var nameRe = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$`)

// SnapshotDir mirrors snapshot_dir().
func SnapshotDir(projectDir string) string {
	return filepath.Join(projectDir, ".shiploom", SnapshotDirname)
}

func checkName(name string) error {
	if !nameRe.MatchString(name) {
		return fmt.Errorf("bad snapshot name %q (letters/digits/._-, max 64)", name)
	}
	return nil
}

// message parity note: Python formats %r of the name (single-quoted for
// these inputs); %q matches for names without quotes/backslashes, which
// the regex above guarantees. Identical output on all valid inputs.

// resolveCommand mirrors _resolve_command().
func resolveCommand(projectDir, command string, timeoutS float64) (string, float64, error) {
	if command != "" {
		return command, timeoutS, nil
	}
	configured := execrun.LoadGateConfig(projectDir)
	cmd, timeout, _, ok := execrun.ResolveTestCommand(projectDir)
	_ = configured
	if !ok {
		// Mirror the ValueError text, including the reason.
		_, _, reason, _ := execrun.ResolveTestCommand(projectDir)
		return "", 0, fmt.Errorf("no command: pass --command or configure gates.test (%s)", reason)
	}
	if timeoutS != 0 {
		return cmd, timeoutS, nil
	}
	return cmd, timeout, nil
}

// Capture mirrors capture(). Returns the entry.
func Capture(projectDir, name, command string, timeoutS float64, actor string) (map[string]any, error) {
	if err := checkName(name); err != nil {
		return nil, err
	}
	command, timeout, err := resolveCommand(projectDir, command, timeoutS)
	if err != nil {
		return nil, err
	}
	if timeout == 0 {
		timeout = DefaultTimeoutS
	}
	result := execrun.RunCommand(projectDir, command, timeout)
	out := result.Tail
	entry := map[string]any{
		"name": name, "command": command, "exit": int64(result.Exit),
		"outputSha": shaOf(out), "tail": tailRunes(out, TailLimit),
		"timeoutS": timeout, "capturedAt": manifest.Utcnow(), "actor": actor,
	}
	dest := SnapshotDir(projectDir)
	if err := os.MkdirAll(dest, 0o755); err != nil {
		return nil, err
	}
	raw, err := jsoncanon.Marshal(entry)
	if err != nil {
		return nil, err
	}
	if err := os.WriteFile(filepath.Join(dest, name+".json"), append(raw, '\n'), 0o644); err != nil {
		return nil, err
	}
	if _, err := auditlog.Append(projectDir, actor, "characterize.capture", name, "-"); err != nil {
		return nil, err
	}
	return entry, nil
}

// load mirrors _load().
func load(projectDir, name string) (map[string]any, error) {
	if err := checkName(name); err != nil {
		return nil, err
	}
	raw, err := os.ReadFile(filepath.Join(SnapshotDir(projectDir), name+".json"))
	if err != nil {
		return nil, fmt.Errorf("no snapshot %q (capture first): %s", name, fsErrText(err))
	}
	doc, err := jsoncanon.Decode(raw)
	if err != nil {
		return nil, fmt.Errorf("no snapshot %q (capture first): %s", name, err)
	}
	obj, ok := doc.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("corrupt snapshot %q (recapture)", name)
	}
	if _, ok := obj["outputSha"]; !ok {
		return nil, fmt.Errorf("corrupt snapshot %q (recapture)", name)
	}
	return obj, nil
}

// ListSnapshots mirrors list_snapshots().
func ListSnapshots(projectDir string) []string {
	dest := SnapshotDir(projectDir)
	items, err := os.ReadDir(dest)
	if err != nil {
		return []string{}
	}
	var out []string
	for _, item := range items {
		if item.IsDir() || filepath.Ext(item.Name()) != ".json" {
			continue
		}
		out = append(out, strings.TrimSuffix(item.Name(), ".json"))
	}
	sort.Strings(out)
	return out
}

// DiffResult mirrors diff()'s return mapping.
type DiffResult struct {
	Name          string
	Changed       bool
	ExitChanged   bool
	OutputChanged bool
	UnifiedDiff   []string
	Current       map[string]any
	Err           string
	HasError      bool
}

// Diff mirrors diff(). Changed behavior is data (warn), not failure.
func Diff(projectDir, name string, timeoutS float64) DiffResult {
	entry, err := load(projectDir, name)
	if err != nil {
		return DiffResult{Name: name, Err: err.Error(), HasError: true}
	}
	var timeout float64
	if timeoutS != 0 {
		timeout = timeoutS
	} else if t, ok := execrun.ToFloat(entry["timeoutS"]); ok && t != 0 {
		timeout = t
	} else {
		timeout = DefaultTimeoutS
	}
	command, _ := entry["command"].(string)
	result := execrun.RunCommand(projectDir, command, timeout)
	out := result.Tail
	exitChanged := !valuesEqual(result.Exit, entry["exit"])
	outputChanged := shaOf(out) != strField(entry, "outputSha")
	var unified []string
	if outputChanged {
		oldTail, _ := entry["tail"].(string)
		unified = unifiedDiff(splitLines(oldTail), splitLines(out))
		if len(unified) > DiffMaxLines {
			unified = unified[:DiffMaxLines]
		}
	}
	// audit best-effort (mirrors try/except pass).
	_, _ = auditlog.Append(projectDir, "system", "characterize.diff", name, "-")
	return DiffResult{Name: name, Changed: exitChanged || outputChanged,
		ExitChanged: exitChanged, OutputChanged: outputChanged, UnifiedDiff: unified,
		Current: map[string]any{"exit": int64(result.Exit), "outputSha": shaOf(out)}}
}

func strField(obj map[string]any, key string) string {
	s, _ := obj[key].(string)
	return s
}

func tailRunes(s string, limit int) string {
	runes := []rune(s)
	if len(runes) <= limit {
		return s
	}
	return string(runes[len(runes)-limit:])
}

func fsErrText(err error) string {
	if pe, ok := err.(*os.PathError); ok {
		return pe.Err.Error()
	}
	return err.Error()
}

func unifiedDiff(oldLines, newLines []string) []string {
	return difflib.UnifiedDiff(oldLines, newLines)
}

func shaOf(text string) string {
	sum := sha256.Sum256([]byte(text))
	return "sha256:" + hex.EncodeToString(sum[:])
}

// valuesEqual mirrors Python !=/== for exit-code comparison.
func valuesEqual(a, b any) bool {
	// Python True == 1 cross-type equality, checked before numerics
	// (numFloat rejects bools, mirroring isinstance int-not-bool).
	if ab, ok := a.(bool); ok {
		if bb, ok := b.(bool); ok {
			return ab == bb
		}
		if bf, ok := numFloat(b); ok {
			return (bf == 1 && ab) || (bf == 0 && !ab)
		}
		return false
	}
	if bb, ok := b.(bool); ok {
		if af, ok := numFloat(a); ok {
			return (af == 1 && bb) || (af == 0 && !bb)
		}
		return false
	}
	af, aok := numFloat(a)
	bf, bok := numFloat(b)
	if aok && bok {
		return af == bf
	}
	if a == nil || b == nil {
		return a == nil && b == nil
	}
	if as, ok := a.(string); ok {
		bs, ok := b.(string)
		return ok && as == bs
	}
	return false
}

func numFloat(v any) (float64, bool) {
	switch t := v.(type) {
	case int64:
		return float64(t), true
	case int:
		return float64(t), true
	case float64:
		return t, true
	case json.Number:
		if f, err := strconv.ParseFloat(string(t), 64); err == nil {
			return f, true
		}
		return 0, false
	default:
		return 0, false
	}
}

func splitLines(text string) []string {
	if text == "" {
		return []string{}
	}
	// Mirror str.splitlines for \n texts (corpus uses \n).
	lines := strings.Split(text, "\n")
	for i := range lines {
		lines[i] = strings.TrimSuffix(lines[i], "\r")
	}
	return lines
}
